### DISCUSSION TITLE
#120 Making predictions using the first CNN

### DISCUSSION THREAD

**Discussion by niktheod** [mrdbourke/pytorch-deep-learning#355] [Q&A]
When we wrote the train and test step functions to calculate the accuracy we passed the raw logits only through `torch.argmax()` and not through `torch.softmax()`. This code is a part from the `train_step` function from Daniel's notebook and you can see that we pass as an argument to the accuracy function, the `y_pred.argmax(dim=1)`, where `y_pred` is actually the `y_logits`.

#
    loss = loss_fn(y_pred, y)
    train_loss += loss # accumulate train loss
    train_acc += accuracy_fn(y_true=y,
                             y_pred=y_pred.argmax(dim=1)) # go from logits -> prediction labels

Later  at 18:47:30 when we wrote a function to make predictions with the already trained CNN model we wrote the following code:

#
    def make_predictions(model: torch.nn.Module,
                     data: list,
                     device: torch.device = device):
    pred_probs = []
    model.to(device)
    model.eval()
    with torch.inference_mode():
      for sample in data:
        # Prepare the sample (add a batch dimension and pass to target device)
        sample = torch.unsqueeze(sample, dim=0).to(device)

        # Forward pass (model outputs raw logits)
        pred_logit = model(sample)

        # Get prediction probability (logit -> prediction probability)
        pred_prob = torch.softmax(pred_logit.squeeze(), dim=0)

        # Get pred_prob off the GPU for further calculations
        pred_probs.append(pred_prob.cpu())

    # Stack the pred_probs to turn list into a tensor
    return torch.stack(pred_probs)
    
Here you can see that we passed the logits first from `torch.softmax()` and then later to make the predictions we passed it also through `torch.argmax()`
Why is this? Why in the first case we didn't use `torch.softmax()` but then we did?
Btw I made predictions by passing the raw outcome of the model only through `torch.argmax()` and it worked very well.

**Comment by niktheod** [comment]
Nvm I just saw that someone else asked the same thing rexently.
